src/AddAtomicMutex.cpp

```diff
@@ -0,0 +1,344 @@
+#include "AddAtomicMutex.h"
+#include "IREquality.h"
+#include "IRMutator.h"
+#include "IROperator.h"
+
+namespace Halide {
+namespace Internal {
+
+using std::set;
+using std::string;
+
+namespace {
+
+class FindLoad : public IRGraphVisitor {
+public:
+    using IRGraphVisitor::visit;
+
+    set<string> symbols;
+    bool found = false;
+
+    void visit(const Load *op) override {
+        if (symbols.find(op->name) != symbols.end()) {
+            found = true;
+        }
+        include(op->predicate);
+        include(op->index);
+    }
+
+    bool find_load(Expr e, const set<string> &symbols) {
+        found = false;
+        this->symbols = symbols;
+        include(e);
+        return found;
+    }
+};
+
+class FindAtomicLetBindings : public IRGraphVisitor {
+public:
+    using IRVisitor::visit;
+
+    FindAtomicLetBindings(const set<string> &store_names)
+        : store_names(store_names) {}
+
+    void visit(const Let *op) override {
+        include(op->value);
+        {
+            ScopedBinding<Expr> bind(let_bindings, op->name, op->value);
+            include(op->body);
+        }
+    }
+
+    void visit(const LetStmt *op) override {
+        include(op->value);
+        {
+            ScopedBinding<Expr> bind(let_bindings, op->name, op->value);
+            include(op->body);
+        }
+    }
+
+    void visit(const Variable *op) override {
+        if (inside_store) {
+            if (let_bindings.contains(op->name)) {
+                Expr e = let_bindings.get(op->name);
+                FindLoad finder;
+                if (finder.find_load(e, store_names)) {
```

Comment:
What happens with:
```
let a = f[x]
let b = 2*a
f[x] = b + 1
```

Reply:
I think you want to maintain a Scope of all the vars that transitively depend on one of the store names, rather than your let_bindings scope

Reply:
Done. I uses `expr_uses_vars` with the let binding scope to check it.